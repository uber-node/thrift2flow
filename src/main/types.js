/*
 * MIT License
 *
 * Copyright (c) 2017 Uber Node.js
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// @flow

import {BaseType, Enum, ListType, MapType, SetType} from 'thriftrw/ast';

export class TypeConverter {
  static primitives = {
    binary: 'Buffer',
    bool: 'boolean',
    byte: 'number',
    i8: 'number',
    i16: 'number',
    i32: 'number',
    i64: 'Buffer',
    double: 'number',
    string: 'string',
    void: 'void'
  };

  static i64Mappings = {
    Long: 'Long',
    Date: 'string'
  };

  transformName: string => string;
  thriftAstDefinitions: Array<any>;

  constructor(transformName: string => string, thriftAstDefinitions: Array<any>) {
    this.transformName = transformName;
    this.thriftAstDefinitions = thriftAstDefinitions;
  }

  annotation(t: BaseType): string {
    const jsType = t.annotations && t.annotations['js.type'];
    // https://github.com/thriftrw/thriftrw-node/blob/8d36b5b83e5d22bf6c28339e3e894eb4926e556f/i64.js#L179
    if (t.baseType === 'i64' && jsType) {
      return TypeConverter.i64Mappings[jsType];
    }
    return '';
  }

  convert = (t: BaseType): string =>
    this.arrayType(t) ||
    this.enumType(t) ||
    this.mapType(t) ||
    this.annotation(t) ||
    TypeConverter.primitives[t.baseType] ||
    this.transformName(t.name);

  enumType = (thriftValueType: BaseType) =>
    this.isEnum(thriftValueType) && this.transformName(thriftValueType.name);

  arrayType = (thriftValueType: BaseType) =>
    (thriftValueType instanceof ListType || thriftValueType instanceof SetType) &&
    `${this.convert(thriftValueType.valueType)}[]`;

  mapType(t: BaseType) {
    if (t instanceof MapType) {
      const ktype = this.convert(t.keyType);
      const vtype = this.convert(t.valueType);
      return `{[${ktype}]: ${vtype}}`;
    }
    return null;
  }

  isEnum(def: BaseType) {
    // Enums export const, not type
    let defName = def.name;
    let defValueType = this.thriftAstDefinitions.find(value => {
      return value.id.name === defName;
    });

    return defValueType && defValueType instanceof Enum;
  }
}
