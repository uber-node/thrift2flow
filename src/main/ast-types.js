// @flow

type Annotations = {|
  'js.type'?: 'Long' | 'Date',
|};
type Primitives =
  | 'i8'
  | 'i16'
  | 'string'
  | 'i32'
  | 'i64'
  | 'double'
  | 'void'
  | 'binary'
  | 'bool'
  | 'byte';
export type BaseType = {|
  type: 'BaseType',
  baseType: Primitives,
  annotations: Annotations,
|};

export type Identifier = {|
  type: 'Identifier',
  name: string,
  annotations: Annotations,
|};

export type Struct = {|
  id: Identifier,
  type: 'Struct',
  fields: Array<Field>,
|};

export type Union = {|
  id: Identifier,
  type: 'Union',
  fields: Array<Field>,
|};

export type Exception = {|
  id: Identifier,
  type: 'Exception',
  fields: Array<Field>,
|};

export type FieldIdentifier = {|
  type: 'FieldIdentifier',
  value: number,
  line: number,
  column: number,
|};
export type Field = {|
  id: FieldIdentifier,
  type: 'Field',
  name: string,
  optional: boolean,
  valueType: Identifier | BaseType,
  required: boolean,
  optional: boolean,
  defaultValue: {||} | null,
  annotations: Annotations,
|};

export type EnumDefinition = {|
  type: 'EnumDefinition',
  id: Identifier,
  fieldType: Identifier | Literal,
  value: Literal,
  annotations: Annotations,
|};

export type Enum = {|
  id: Identifier,
  type: 'Enum',
  definitions: Array<EnumDefinition>,
  annotations: Annotations,
|};

export type Typedef = {|
  id: Identifier,
  type: 'Typedef',
  valueType: Identifier | BaseType,
|};

export type FunctionDefinition = {|
  id: Identifier,
  fields: Array<Field>,
  returns: Identifier | BaseType,
|};

export type Service = {|
  id: Identifier,
  type: 'Service',
  functions: Array<FunctionDefinition>,
|};

export type Const = {|
  id: Identifier,
  type: 'Const',
  fieldType: Identifier | Literal | BaseType,
  value: ConstList | ConstEntry | ConstMap | Literal,
|};

export type Literal = {|
  id: Identifier,
  type: 'Literal',
  value: string | number,
|};

export type ConstList = {|
  id: Identifier,
  type: 'ConstList',
  values: Array<Identifier | Literal>,
|};

export type ConstEntry = {|
  id: Identifier,
  type: 'ConstEntry',
  key: Literal | Identifier,
  value: Literal | Identifier | ConstMap,
|};

export type List = {|
  id: Identifier,
  type: 'List',
  valueType: Identifier | BaseType,
|};

export type Map = {|
  id: Identifier,
  type: 'Map',
  keyType: Identifier | BaseType,
  valueType: Identifier | BaseType,
|};

export type Set = {|
  id: Identifier,
  type: 'Set',
  valueType: Identifier | BaseType,
|};

export type ConstMap = {|
  id: Identifier,
  type: 'ConstMap',
  entries: Array<ConstEntry>,
|};

export type Definition =
  | Struct
  | Exception
  | Union
  | Enum
  | EnumDefinition
  | Typedef
  | Service
  | Const;

export type AstNode =
  | Definition
  | Identifier
  | Literal
  | BaseType
  | List
  | Map
  | Set
  | Field
  | EnumDefinition
  | Union
  | ConstEntry
  | ConstMap;

export type Ast = {|
  definitions: Array<Definition>,
  headers: $ReadOnlyArray<{|
    type: 'Include',
    // ie., foo.thrift
    id: string,
    namespace: string | null,
    line: number,
    column: number,
  |}>,
|};
