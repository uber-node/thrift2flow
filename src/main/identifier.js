// @flow
const reservedTypes = ['any', 'mixed', 'number'];

export function id(s: string): string {
  const split = s.split('.');
  if (reservedTypes.includes(s) || reservedTypes.includes(split[0])) {
    return `_${s}`;
  }
  return s;
}
