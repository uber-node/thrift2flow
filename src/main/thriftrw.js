// @flow

import {Thrift} from 'thriftrw';
import type {Ast} from './ast-types';
// This file is a type wrapper to thriftrw

export default function(options: {|
  strict: boolean,
  allowFilesystemAccess: boolean,
  allowOptionalArguments: boolean,
  entryPoint: string,
|}): {|
  asts: {[filename: string]: Ast},
  filename: string,
  idls: {[filename: string]: {||}},
|} {
  return new Thrift(options);
}
