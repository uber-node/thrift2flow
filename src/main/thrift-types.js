/*
 * MIT License
 *
 * Copyright (c) 2017 Uber Node.js
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// @flow

import {id} from './identifier';
import type {AstNode} from './ast-types';

const primitives = {
  binary: 'Buffer',
  bool: 'boolean',
  byte: 'number',
  i8: 'number',
  i16: 'number',
  i32: 'number',
  i64: 'Buffer',
  double: 'number',
  string: 'string',
  void: 'void',
};

const i64Mappings = {
  Long: 'thrift2flow$Long',
  Date: 'string',
};

export class TypeConverter {
  thriftAstDefinitions: Array<any>;
  identifiersTable: $ReadOnly<{[key: string]: AstNode}>;

  constructor(
    thriftAstDefinitions: Array<any>,
    identifiersTable: $ReadOnly<{[key: string]: AstNode}>
  ) {
    this.thriftAstDefinitions = thriftAstDefinitions;
    this.identifiersTable = identifiersTable;
  }

  baseType(t: AstNode): string | void {
    if (t.type !== 'BaseType') {
      return undefined;
    }
    if (t.baseType === 'i64') {
      const jsType = t.annotations['js.type'];
      if (jsType !== undefined) {
        return i64Mappings[jsType];
      }
    }
    return primitives[t.baseType];
  }

  convert = (t: AstNode): string => {
    if (!t) {
      throw new Error(`Assertion failed. t is not defined.`);
    }
    let type =
      this.arrayType(t) ||
      this.mapType(t) ||
      this.enumType(t) ||
      this.baseType(t);
    if (type) {
      return type;
    }
    if (t.type === 'Identifier') {
      type = id(t.name);
    }
    if (type) {
      return type;
    }
    throw new Error(`Unhandled convertion for node ${JSON.stringify(t)}`);
  };

  enumType = (thriftValueType: AstNode): string | void => {
    if (this.isEnum(thriftValueType)) {
      // Enums are values, not types. To refer to the type,
      // we use $Values<...>.
      if (thriftValueType.type !== 'Identifier') {
        throw new Error(
          'Assertion failure. Enum reference has to be an identifier'
        );
      }
      return `$Values<typeof ${thriftValueType.name}>`;
    }
    return undefined;
  };

  arrayType = (node: AstNode) =>
    ((node.type === 'List' || node.type === 'Set') &&
      `${this.convert(node.valueType)}[]`) ||
    undefined;

  mapType(t: AstNode): string | void {
    if (t.type === 'Map') {
      const keyType = this.convert(t.keyType);
      const valueType = this.convert(t.valueType);
      return `{[${keyType}]: ${valueType}}`;
    }
    return undefined;
  }

  isEnum(def: AstNode) {
    // Enums export const, not type
    if (!def.name) {
      return undefined;
    }
    return this.identifiersTable[def.name].type === 'Enum';
  }
}
