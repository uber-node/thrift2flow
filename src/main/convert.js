/*
 * MIT License
 *
 * Copyright (c) 2017 Uber Node.js
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// @flow

import {Thrift} from 'thriftrw';
import {TypeConverter} from './types';
import prettier from 'prettier';
import path from 'path';

import type {Base} from 'bufrw';
import type {
  Struct,
  Field,
  Enum,
  Typedef,
  FunctionDefinition,
  Service,
  Const,
} from 'thriftrw/ast';

const thriftOptions = {
  strict: false,
  allowFilesystemAccess: true,
  allowOptionalArguments: true,
};

export class ThriftFileConverter {
  thriftPath: string;
  thrift: Thrift;
  types: TypeConverter;
  transformName: string => string;
  withsource: boolean;
  ast: any;
  thriftAstDefinitions: Array<any>;

  constructor(
    thriftPath: string,
    transformName: string => string,
    withsource: boolean
  ) {
    this.thriftPath = path.resolve(thriftPath);
    this.thrift = new Thrift({...thriftOptions, entryPoint: thriftPath});
    this.ast = this.thrift.asts[this.thrift.filename];
    this.thriftAstDefinitions = this.ast.definitions;
    this.transformName = transformName;
    this.types = new TypeConverter(transformName, this.thriftAstDefinitions);
    this.withsource = withsource;
  }

  generateFlowFile = () =>
    prettier.format(
      [
        '// @flow',
        `// Generated by thrift2flow at ${new Date().toString()}${
          this.withsource ? `\n// Source: ${this.thriftPath}` : ''
        }`,
        this.generateImports(),
        ...this.thriftAstDefinitions.map(this.convertDefinitionToCode),
      ]
        .filter(Boolean)
        .join('\n\n'),
      {parser: 'flow'}
    );

  convertDefinitionToCode = (def: any) => {
    switch (def.type) {
      case 'Struct':
      case 'Exception':
        return this.generateStruct(def);
      case 'Union':
        return this.generateUnion(def);
      case 'Enum':
        return this.generateEnum(def);
      case 'Typedef':
        return this.generateTypedef(def);
      case 'Service':
        return this.generateService(def);
      case 'Const':
        return this.generateConst(def);
      default:
        console.warn(
          `${path.basename(this.thriftPath)}: Skipping ${def.type} ${
            def.id ? def.id.name : '?'
          }`
        );
        return null;
    }
  };

  generateService = (def: Service) =>
    `export type ${this.transformName(def.id.name)} = {\n${def.functions
      .map(this.generateFunction)
      .join(',')}};`;

  generateFunction = (fn: FunctionDefinition) =>
    `${fn.id.name}: (${
      fn.fields.length ? this.generateStructContents([...fn.fields]) : ''
    }) => ${this.types.convert(fn.returns)}`;

  generateTypedef = (def: Typedef) =>
    `export type ${this.transformName(def.id.name)} = ${this.types.convert(
      def.valueType
    )};`;

  generateEnumUnion = (def: Enum) => {
    return def.definitions.map((d, index) => `"${d.id.name}"`).join(' | ');
  };

  generateEnumType = (def: Enum) => {
    return `export type ${this.transformName(
      def.id.name
    )} = ${this.generateEnumUnion(def)};`;
  };

  generateEnumMap = (def: Enum) => {
    const header = '{';
    const values = def.definitions
      .map(
        (d, index) => `  "${d.id.name}": ${d.value ? d.value.value : index},`
      )
      .join('\n');
    const footer = '}';

    const mapDefinition = [header, values, footer].join('\n');
    return `export const ${def.id.name}ValueMap = ${mapDefinition};`;
  };

  generateEnum = (def: Enum) => {
    return `${this.generateEnumType(def)}\n${this.generateEnumMap(def)}`;
  };

  generateConst = (def: Const) => {
    let value;
    if (def.value.type === 'ConstList') {
      value = `[${def.value.values
        .map(val => {
          if (val.type === 'Identifier') {
            return val.name;
          }
          if (typeof val.value === 'string') {
            return `'${val.value}'`;
          }
          return val.value;
        })
        .join(',')}]`;
    } else {
      value =
        typeof def.value.value === 'string'
          ? `'${def.value.value}'`
          : def.value.value;
    }
    return `export const ${def.id.name}: ${this.types.convert(
      def.fieldType
    )} = ${value};`;
  };

  generateStruct = ({id: {name}, fields}: Struct) =>
    `export type ${this.transformName(name)} = ${this.generateStructContents(
      fields
    )};`;

  generateStructContents = (fields: Object) =>
    `{|${Object.values(fields)
      .map(
        (f: Base) =>
          `${f.name}${this.isOptional(f) ? '?' : ''}: ${this.types.convert(
            f.valueType
          )};`
      )
      .join('\n')}|}`;

  generateUnion = ({id: {name}, fields}: Struct) =>
    `export type ${this.transformName(name)} = ${this.generateUnionContents(
      fields
    )};`;

  generateUnionContents = (fields: Object) => {
    if (!fields.length) {
      return '{||}';
    }
    return Object.values(fields)
      .map((f: Base) => {
        return `{|${f.name}: ${this.types.convert(f.valueType)}|}`;
      })
      .join(' | ');
  };

  isOptional = (field: Field) => field.optional;

  generateImports = () => {
    const includes = this.ast.headers.filter(f => f.type === 'Include');
    const relativePaths = includes
      .map(i => path.parse(i.id))
      .map(parsed => path.join(parsed.dir, parsed.name))
      .map(p => (p.startsWith('.') ? p : `./${p}`));
    const generatedImports = relativePaths.map((relpath, index) => {
      let baseName = path.basename(relpath);
      let hasConflictingImport = true;
      while (hasConflictingImport) {
        hasConflictingImport = relativePaths.some((rel, nextIndex) => {
          if (nextIndex > index && path.basename(rel) === baseName) {
            return true;
          }
          return false;
        });
        if (hasConflictingImport) {
          baseName = `_${baseName}`;
        }
      }
      return `import * as ${baseName} from '${relpath}';`;
    });

    if (this.isLongDefined()) {
      generatedImports.push("import Long from 'long'");
    }
    return generatedImports.join('\n');
  };
  getImportAbsPaths = () =>
    Object.keys(this.thrift.idls).map(p => path.resolve(p));

  isLongDefined = () => {
    for (const astNode of this.thriftAstDefinitions) {
      if (astNode.type === 'Struct') {
        for (const field of astNode.fields) {
          if (field.valueType == null || field.valueType.annotations == null) {
            continue;
          }

          if (field.valueType.annotations['js.type'] === 'Long') {
            return true;
          }
        }
      } else if (astNode.type === 'Typedef') {
        if (
          astNode.valueType == null ||
          astNode.valueType.annotations == null
        ) {
          continue;
        }

        if (astNode.valueType.annotations['js.type'] === 'Long') {
          return true;
        }
      }
    }

    return false;
  };
}
