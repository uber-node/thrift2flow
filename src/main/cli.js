#!/usr/bin/env node
/*
 * MIT License
 *
 * Copyright (c) 2017 Uber Node.js
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* eslint-disable no-process-exit */
// @flow

import 'source-map-support/register';
import 'babel-polyfill';
import path from 'path';
import fs from 'fs';
// flowlint-next-line untyped-import:off
import yargs from 'yargs';
// flowlint-next-line untyped-import:off
import mkdirp from 'mkdirp';
import convert from './index';

const argv = yargs
  .usage('Usage: $0 [options] <thrift files..>')
  .option('path', {
    describe:
      'common path used for generated code. In most projects this is likely just `idl`.',
  })
  .option('outputDir', {
    describe: 'The root output directory for the generated code.',
    default: 'flow-output',
  })
  .option('withsource', {
    describe: 'prepend the source path of the thrift file',
    default: false,
  })
  .help('h')
  .alias('h', 'help').argv;

const thriftPaths = argv._;

if (!thriftPaths.length) {
  yargs.showHelp();
  process.exit(1);
}

const options = {
  withSource: argv.withsource,
  commonPath: path.resolve(process.cwd(), argv.path),
  outputDir: argv.outputDir,
};

const allOutput = convert(thriftPaths, options);

for (const jsFilename in allOutput) {
  mkdirp(path.dirname(jsFilename), () =>
    fs.writeFile(jsFilename, allOutput[jsFilename], () =>
      console.log(`Wrote ${jsFilename}`)
    )
  );
}
