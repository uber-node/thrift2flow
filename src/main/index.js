#!/usr/bin/env node
/*
 * MIT License
 *
 * Copyright (c) 2017 Uber Node.js
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* eslint-disable no-process-exit */
// @flow

import 'source-map-support/register';

import 'babel-polyfill';
import path from 'path';
import fs from 'fs';
import yargs from 'yargs';

import commonPathPrefix from 'common-path-prefix';
import mkdirp from 'mkdirp';
import {ThriftFileConverter} from './convert';

const argv = yargs
  .usage('Usage: $0 [options] <thrift files..>')
  .option('suffix', {
    describe: 'appended to generated type names',
    default: 'Type',
  })
  .option('path', {
    describe: 'Path used for generated code',
  })
  .option('withsource', {
    describe: 'prepend the source path of the thrift file',
    default: false,
  })
  .help('h')
  .alias('h', 'help').argv;

const thriftPaths = argv._;

if (!thriftPaths.length) {
  yargs.showHelp();
  process.exit(1);
}

const allOutput = {};

for (const thriftPath of thriftPaths) {
  const converter = new ThriftFileConverter(
    thriftPath,
    name => name + argv.suffix,
    argv.withsource
  );
  converter
    .getImportAbsPaths()
    .filter(p => thriftPaths.indexOf(p) === -1)
    .forEach(p => thriftPaths.push(p));
  allOutput[converter.thriftPath] = converter.generateFlowFile();
}

const root = argv.path || commonPathPrefix(Object.keys(allOutput));

for (const thriftPath in allOutput) {
  const relativeThriftPath = path.dirname(path.relative(root, thriftPath));
  const jsFilename = path.resolve(
    'flow-output',
    relativeThriftPath,
    `${path.basename(thriftPath, '.thrift')}.js`
  );
  mkdirp(path.dirname(jsFilename), () =>
    fs.writeFile(jsFilename, allOutput[thriftPath], () =>
      console.log(`Wrote ${jsFilename}`)
    )
  );
}
