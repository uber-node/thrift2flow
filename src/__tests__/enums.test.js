// @flow
/*
 * MIT License
 *
 * Copyright (c) 2017 Uber Node.js
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import {flowResultTest} from '../test-util';
import {ThriftFileConverter} from '../main/convert';
import {Thrift} from 'thriftrw';

test('thriftrw enums work in map constants', () => {
  const fixturePath = 'src/__tests__/fixtures/enums-as-map-keys.thrift';
  const thrift = new Thrift({
    entryPoint: fixturePath,
    allowFilesystemAccess: true
  });
  expect(thrift.THE_ENUM_PROP.DEFAULT).toEqual('DEFAULT');
  const converter = new ThriftFileConverter(fixturePath, false);
  expect(converter.generateFlowFile()).toMatchInlineSnapshot(`
"// @flow

export const THE_ENUM_PROP: $ReadOnly<{|
  DEFAULT: \\"DEFAULT\\",
  LOW: \\"LOW\\",
  HIGH: \\"HIGH\\"
|}> = Object.freeze({
  DEFAULT: \\"DEFAULT\\",
  LOW: \\"LOW\\",
  HIGH: \\"HIGH\\"
});

export const THE_STRING_MAP: $ReadOnly<{|
  \\"0\\": string,
  \\"1\\": string,
  \\"2\\": string
|}> = {
  \\"0\\": \\"Some default string\\",
  \\"1\\": \\"Some low string\\",
  \\"2\\": \\"Some high string\\"
};

export const THE_STRING_KEY_MAP: $ReadOnly<{|
  DEFAULT: string,
  LOW: string,
  HIGH: string
|}> = {
  [THE_ENUM_PROP.DEFAULT]: \\"some other default\\",
  [THE_ENUM_PROP.LOW]: \\"some other low\\",
  [THE_ENUM_PROP.HIGH]: \\"some other high\\"
};
"
`);
});

test('thriftrw enums are strings not numbers', () => {
  const thrift = new Thrift({
    entryPoint: 'src/__tests__/fixtures/my-enum.thrift',
    allowFilesystemAccess: true
  });
  expect(thrift.MyEnum.OK).toEqual('OK');
  expect(thrift.MyEnum.ERROR).toEqual('ERROR');
});

test('enum to JS', () => {
  const converter = new ThriftFileConverter('src/__tests__/fixtures/my-enum.thrift', false);
  const jsContent = converter.generateFlowFile();
  expect(jsContent).toMatchSnapshot();
});

test('enums work with typedefs', () => {
  const converter = new ThriftFileConverter(
    'src/__tests__/fixtures/my-enum-with-typedef.thrift',
    false
  );
  const jsContent = converter.generateFlowFile();
  expect(jsContent).toMatchSnapshot();
});

test('typedefs of enums can be referenced from structs', () => {
  const converter = new ThriftFileConverter(
    `src/__tests__/fixtures/enum-typedef-struct.thrift`,
    false
  );
  expect(converter.generateFlowFile()).toMatchInlineSnapshot(`
"// @flow

export const EnumTypedef: $ReadOnly<{|
  OK: \\"OK\\",
  ERROR: \\"ERROR\\"
|}> = Object.freeze({
  OK: \\"OK\\",
  ERROR: \\"ERROR\\"
});

export const MyEnum: $ReadOnly<{|
  OK: \\"OK\\",
  ERROR: \\"ERROR\\"
|}> = Object.freeze({
  OK: \\"OK\\",
  ERROR: \\"ERROR\\"
});

export type MyStruct = {|
  f_MyEnum: $Values<typeof MyEnum>,
  f_EnumTypedef: $Values<typeof EnumTypedef>
|};
"
`);
});

test('enums with typedefs', done => {
  flowResultTest(
    {
      'types.thrift': `
typedef MyEnum EnumTypedef

enum MyEnum {
  OK = 1
  ERROR = 2
}

struct MyStruct {
  1: MyEnum f_MyEnum
  2: EnumTypedef f_EnumTypedef
}
`,
      'index.js': `
// @flow
import { MyEnum, type MyStruct, EnumTypedef } from './types';

const ok: $Values<typeof MyEnum> = 'OK';
const error: $Values<typeof MyEnum> = 'ERROR';

const struct: MyStruct = {
  f_MyEnum: ok,
  f_EnumTypedef: error,
}

const okFromMap: 'OK' = MyEnum.OK;
const errorFromMap: 'ERROR' = MyEnum.ERROR;

const t: $Values<typeof EnumTypedef> = ok;
`
    },
    (r: FlowResult) => {
      expect(r.errors).toEqual([]);
      done();
    }
  );
});

test('enums with errors', done => {
  flowResultTest(
    {
      // language=thrift
      'types.thrift': `
typedef MyEnum EnumTypedef

enum MyEnum {
  OK = 1
  ERROR = 2
}

struct MyStruct {
  1: MyEnum f_MyEnum
  2: EnumTypedef f_EnumTypedef
}
`,
      // language=JavaScript
      'index.js': `
// @flow
import {type MyStruct, EnumTypedef, MyEnum} from './types';

const ok: $Values<typeof MyEnum> = 'NOT CORRECT';
const error: $Values<typeof MyEnum> = null;

const struct: MyStruct = {
  f_MyEnum: 'NOT CORRECT',
  f_EnumTypedef: null,
}

const t: $Values<typeof EnumTypedef> = 'NOT CORRECT';
`
    },
    (r: FlowResult) => {
      expect(r.errors.length).toEqual(5);
      done();
    }
  );
});
