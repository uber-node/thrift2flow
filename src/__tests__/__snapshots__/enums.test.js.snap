// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enum to JS 1`] = `
"// @flow

export type MyEnum = \\"OK\\" | \\"ERROR\\";
export const MyEnumValueMap = {
  OK: 1,
  ERROR: 2
};
"
`;
